sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TTLState.scala

```
@@ -0,0 +1,184 @@
+    RangeKeyScanStateEncoderSpec(TTL_KEY_ROW_SCHEMA, 1), isInternal = true)
+
+  def upsertTTLForStateKey(
+      expirationMs: Long,
+      groupingKey: Array[Byte]): Unit = {
+    val encodedTtlKey = ttlKeyEncoder(InternalRow(expirationMs, groupingKey))
+    store.put(encodedTtlKey, EMPTY_ROW, ttlColumnFamilyName)
+  }
+
+  override def clearExpiredState(): Unit = {
+    val iterator = store.iterator(ttlColumnFamilyName)
+
+    iterator.takeWhile { kv =>
+      val expirationMs = kv.key.getLong(0)
+      StateTTL.isExpired(ttlMode, expirationMs,
+        batchTimestampMs, eventTimeWatermarkMs)
+    }.foreach { kv =>
+      val groupingKey = kv.key.getBinary(1)
+      state.clearIfExpired(groupingKey)
+      store.remove(kv.key, ttlColumnFamilyName)
+    }
+  }
+
+  private[sql] def setStateVariable(
+      state: StateVariableWithTTLSupport): Unit = {
+    this.state = state
+  }
+
+  private[sql] def iterator(): Iterator[SingleKeyTTLRow] = {
+    val ttlIterator = store.iterator(ttlColumnFamilyName)
+
+    new Iterator[SingleKeyTTLRow] {
+      override def hasNext: Boolean = ttlIterator.hasNext
+
+      override def next(): SingleKeyTTLRow = {
+        val kv = ttlIterator.next()
+        SingleKeyTTLRow(
+          expirationMs = kv.key.getLong(0),
+          groupingKey = kv.key.getBinary(1)
+        )
+      }
+    }
+  }
+}
+
+/**
+ * Helper methods for user State TTL.
+ */
+object StateTTL {
+  def calculateExpirationTimeForDuration(
+      ttlMode: TTLMode,
+      ttlDuration: Duration,
+      batchTimestampMs: Option[Long],
+      eventTimeWatermarkMs: Option[Long]): Long = {
+    if (ttlMode == TTLMode.ProcessingTimeTTL()) {
+      batchTimestampMs.get + ttlDuration.toMillis
+    } else if (ttlMode == TTLMode.EventTimeTTL()) {
+      eventTimeWatermarkMs.get + ttlDuration.toMillis
+    } else {
+      throw new IllegalStateException(s"cannot calculate expiration time for" +
```

Could we throw a NERF error if required ?

---

This code path is no longer needed.